src/test/java/org/opendatakit/common/persistence/NewQueryResultTest.java

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright (C) 2011 University of Washington
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.opendatakit.common.persistence;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.opendatakit.common.persistence.Query.Direction.ASCENDING;
+import static org.opendatakit.common.persistence.TestTable.booleanField;
+import static org.opendatakit.common.persistence.TestTable.dateField;
+import static org.opendatakit.common.persistence.TestTable.doubleField;
+import static org.opendatakit.common.persistence.TestTable.integerField;
+import static org.opendatakit.common.persistence.TestTable.stringField;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.opendatakit.common.persistence.exception.ODKDatastoreException;
+import org.opendatakit.common.persistence.exception.ODKEntityPersistException;
+import org.opendatakit.common.persistence.exception.ODKOverQuotaException;
+import org.opendatakit.common.security.User;
+import org.opendatakit.common.web.CallingContext;
+import org.opendatakit.common.web.TestContextFactory;
+
+
+public class NewQueryResultTest {
+  private static final int SET_SIZE = 3;
+  private static final String[] STRINGS = new String[]{"A", "B", "C"};
+  private static final double[] DOUBLES = new double[]{0.9, 1.9, 2.9};
+  private static final String[] DATES = new String[]{"2009-01-30", "2009-02-30", "2009-03-30"};
+  private static final Boolean[] BOOLEANS = new Boolean[]{true, false, null};
+  private static TestRow[] TEST_ROWS = new TestRow[STRINGS.length * SET_SIZE * DOUBLES.length * DATES.length * BOOLEANS.length];
+
+  static {
+    int i = 0;
+    for (String someString : STRINGS)
+      for (int someInteger = 0; someInteger < SET_SIZE; someInteger++)
+        for (Double someDouble : DOUBLES)
+          for (String someDate : DATES)
+            for (Boolean someBoolean : BOOLEANS)
+              TEST_ROWS[i++] = new TestRow(someString, someInteger, someDouble, someDate, someBoolean);

```

Comment:
This is a cool approach to generating data.

Reply:
That's a cartesian product with Vavr.io...